### ISSUE TITLE
Help me understand

### ISSUE THREAD

**Issue by thevickypedia** [azimjohn/jprq.live#14]
Can someone explain what's happening at `jprq.io`?
Curl shows the head: `nginx/1.18.0 (Ubuntu)`
Curious to know how the connection is being tunneled

**Comment by azimjohn** [comment]
Hello, there's Nginx behind the jprq.io server that is used for securing with SSL.
You can find the backend of tunneling here https://github.com/azimjohn/jprq

In the near future, I am going to remove the need for an Nginx server and run the jprq server with SSL certificates

**Comment by thevickypedia** [comment]
@azimjohn Thanks for your response. Is there a README, Wiki or something similar (even a YT tutorial) discussing the implementation details of the server? If so, can you please link it to me?

Also, serving certs is the only task performed by the server? Because, I see that the websocket connection to the uri returns a token which is used to send websocket messages

**Comment by azimjohn** [comment]
Unfortunately, there's not enough documentation/videos.

About the way JPRQ HTTP tunnels work, I think it wasn't quite the right solution and for that reason, HTTP-long polling and WebSocket protocols fail to work. I have recently worked on JPRQ TCP tunneling and it helped me make progress on building the HTTP tunneling the right way. After I will have implemented the more sophisticated solution, I will try to document as much as I can.

For now, you can read this article https://dev.to/azimjohn/how-i-built-ngrok-alternative-3n0g
 
In these GitHub issues, I have tried to answer how to self-host jprq:
 - https://github.com/azimjohn/jprq/issues/58
 - https://github.com/azimjohn/jprq-python-client/issues/1

**Comment by thevickypedia** [comment]
Thank you so much, that helps

**Comment by thevickypedia** [comment]
@azimjohn As an effort to create similar solution for tunneling, I came across an issue where my router's port forwarding doesn't work regardless of the port number. I'm unable to access anything via the public IP address. I'm still investigating why my router doesn't forward properly, but I'm curious how jprq solution works around this problem.

**Comment by azimjohn** [comment]
For every outside client connection, the JPRQ client makes 2 connections, one to your server and another to JPRQ Server, and binds one's input to the other's output and one's output to the other's input. And JPRQ server will also bind outside client's connection to the JPRQ client's connection the same way. And it works!

What just happened? JPRQ client made two-way connections with the local server and the JPRQ server.
Note that routers/firewalls cannot expose SERVERS by default. JPRQ client was communicating between the local server and JPRQ server AS A CLIENT and routers/firewalls do not usually limit clients. 

In a nutshell, that's how it works. I am planning to make an explanatory video on this sometime soon.

**Comment by thevickypedia** [comment]
That's a good explanation there. Thanks for your patience.
I'm using caddy and dynamic DNS for my solution, but I'm unable to get around the port forwarding part, because I'm not directly connected to the internet but to another subnet of my ISP 🤦🏻 
I think making two-way connections is where I will be heading next, so I'm looking forward to your explanatory video.
